crypto/keyring_streaming.go

```diff
@@ -0,0 +1,297 @@
+	if !res.isClosed {
+		return nil, errors.New("gopenpgp: can't access key packet until the message writer has been closed")
+	}
+	return res.keyPacket, nil
+}
+
+// EncryptSplitStream is used to encrypt data as a stream.
+// It takes a writer for the Symmetrically Encrypted Data Packet
+// (https://datatracker.ietf.org/doc/html/rfc4880#section-5.7)
+// and returns a writer for the plaintext data and the key packet.
+// If signKeyRing is not nil, it is used to do an embedded signature.
+func (keyRing *KeyRing) EncryptSplitStream(
+	dataPacketWriter Writer,
+	plainMessageMetadata *PlainMessageMetadata,
+	signKeyRing *KeyRing,
+) (*EncryptSplitResult, error) {
+	config := &packet.Config{DefaultCipher: packet.CipherAES256, Time: getTimeGenerator()}
+
+	if plainMessageMetadata == nil {
+		// Use sensible default metadata
+		plainMessageMetadata = &PlainMessageMetadata{
+			IsBinary: true,
+			Filename: "",
+			ModTime:  GetUnixTime(),
+		}
+	}
+
+	hints := &openpgp.FileHints{
+		FileName: plainMessageMetadata.Filename,
+		IsBinary: plainMessageMetadata.IsBinary,
+		ModTime:  time.Unix(plainMessageMetadata.ModTime, 0),
+	}
+
+	var keyPacketBuf bytes.Buffer
+	plainMessageWriter, err := asymmetricEncryptStream(hints, &keyPacketBuf, dataPacketWriter, keyRing, signKeyRing, config)
+	if err != nil {
+		return nil, err
+	}
+	return &EncryptSplitResult{
+		keyPacketBuf:       &keyPacketBuf,
+		plainMessageWriter: plainMessageWriter,
+	}, nil
+}
+
+// PlainMessageReader is used to wrap the data of the decrypted plain message.
+// It can be used to read the decrypted data and verify the embedded signature.
+type PlainMessageReader struct {
+	details       *openpgp.MessageDetails
+	verifyKeyRing *KeyRing
+	verifyTime    int64
+	readAll       bool
+}
+
+// GetMetadata returns the metadata of the decrypted message.
+func (msg *PlainMessageReader) GetMetadata() *PlainMessageMetadata {
+	return &PlainMessageMetadata{
+		Filename: msg.details.LiteralData.FileName,
+		IsBinary: msg.details.LiteralData.IsBinary,
+		ModTime:  int64(msg.details.LiteralData.Time),
+	}
+}
+
+// Read is used to access the message decrypted data.
+// Makes PlainMessageReader implement the Reader interface.
+func (msg *PlainMessageReader) Read(b []byte) (n int, err error) {
+	n, err = msg.details.UnverifiedBody.Read(b)
+	if errors.Is(err, io.EOF) {
+		msg.readAll = true
+	}
+	return
+}
+
+// VerifySignature is used to verify that the signature is valid.
+// This method needs to be called once all the data has been read.
+// It will return an error if the signature is invalid
+// or if the message hasn't been read entirely.
+func (msg *PlainMessageReader) VerifySignature() (err error) {
+	if !msg.readAll {
+		return errors.New("gopenpg: can't verify the signature until the message reader has been read entirely")
+	}
+	if msg.verifyKeyRing != nil {
+		processSignatureExpiration(msg.details, msg.verifyTime)
+		err = verifyDetailsSignature(msg.details, msg.verifyKeyRing)
+	}
```

Comment:
Hmm, I think here we should error if there is no `verifyKeyRing`. Silently "passing" if there is no verifying key can create weird side effects

Reply:
Agreed, I added an error